Improve new empty item creation by using clone when possible

igorwulff opened the pull request:
This patch improves creation time of new objects by using clone whenever possible. I'm still testing this, but would like to get feedback on this. I've been adding a timer that compared the old and new result in the same transaction, when navigating through the site there are some gains to be found. Especially when cached are more cold (like block_html), admin areas like category view show the biggest gains:

I can track per transaction type, these are just logged results for now from across frontend and admin area:
TIMER:-19.572875 ms COUNTER: 2741
TIMER:-1.616937 ms COUNTER: 16
TIMER:-7.571272 ms COUNTER: 1506
TIMER:0 ms COUNTER: 0
TIMER:-1.572855 ms COUNTER: 16
TIMER:-23.672388 ms COUNTER: 801
TIMER:-3.223335 ms COUNTER: 11
TIMER:-13.592401 ms COUNTER: 75
TIMER:-2.116431 ms COUNTER: 11
TIMER:-7.662507 ms COUNTER: 74
TIMER:-4.198262 ms COUNTER: 49
TIMER:-72.5475 ms COUNTER: 5855
TIMER:-4.133812 ms COUNTER: 11
TIMER:-4.779091 ms COUNTER: 49
TIMER:-2.301819 ms COUNTER: 11
TIMER:-8.377065 ms COUNTER: 813
TIMER:-6.833618 ms COUNTER: 813
TIMER:-2.434347 ms COUNTER: 11
TIMER:-3.818474 ms COUNTER: 49
TIMER:-10.169898 ms COUNTER: 814
TIMER:-3.050243 ms COUNTER: 11
TIMER:-3.200632 ms COUNTER: 49
TIMER:-5.976394 ms COUNTER: 813
TIMER:-2.748932 ms COUNTER: 11
TIMER:-3.96872 ms COUNTER: 49
TIMER:-106.360966 ms COUNTER: 5855
TIMER:-3.895319 ms COUNTER: 11
TIMER:-3.334799 ms COUNTER: 49
TIMER:-2.288225 ms COUNTER: 11
TIMER:-60.319299 ms COUNTER: 5855
TIMER:-3.454251 ms COUNTER: 11
TIMER:-3.306505 ms COUNTER: 49
TIMER:-2.563021 ms COUNTER: 11
TIMER:-88.340151 ms COUNTER: 5856
TIMER:-3.875317 ms COUNTER: 11
TIMER:-4.289538 ms COUNTER: 49
TIMER:-2.396556 ms COUNTER: 11
TIMER:-8.741497 ms COUNTER: 813
TIMER:-2.273909 ms COUNTER: 11
TIMER:-3.113306 ms COUNTER: 49
TIMER:-9.474795 ms COUNTER: 674
TIMER:-4.063281 ms COUNTER: 49

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

igorwulff committed:
Improve new empty item creation by using clone when possible

This patch improves creation time of new objects by using clone whenever possible. I'm still testing this, but would like to get feedback on this:

Old version with PLP with all caches enabled:
COUNTER:129
Timer:17.240343 ms

Old version with PLP empty caches:
COUNTER:4470
Timer: 381.315609 ms

Old version with PLP no block cache:
COUNTER:2398
Timer:202.855686 ms

New version with PLP no block cache:
COUNTER:2398
Timer: 2.540631 ms

    
Old version Category Admin page:
COUNTER:5919
Timer: 1423.777139 ms

New version Category Admin page:
COUNTER:5919
Timer: 8.040596ms

ihor-sviziev commented:
@igorwulff I don't think it's the right thing since clone might cause copying data from an existing object to new one, while earlier, it was always creating a new one.

igorwulff commented:
> @igorwulff I don't think it's the right thing since clone might cause copying data from an existing object to new one, while earlier, it was always creating a new one.

You are right. I will think about possible alternative ways to see if and how this can be done.

igorwulff committed:
Update collection for a local instance cache

Update collection for a local instance cache which we can use to clone new empty objects.

igorwulff committed:
Reverted invalid change to docblock

Den4ik commented on lib/internal/Magento/Framework/Data/Collection.php:
```diff
@@ -611,6 +616,13 @@ public function setItemObjectClass($className)
             throw new \InvalidArgumentException($className . ' does not extend \Magento\Framework\DataObject');
         }
+        
+        if ($this->newEmptyItem !== null
+            && $this->_itemObjectClass !== $className
+        ) {
+            $this->newEmptyItem === null;
+        }
```
I think it's redundant code because this is mean that one collection could have multiple items with different classes.

igorwulff replied:
@Den4ik  I've removed the redundant code.

igorwulff committed:
Revert changes to setItemObjectClass method

Revert changes to setItemObjectClass method and removed the additional checks in there that should not be needed.

igorwulff commented:
@magento run all tests

igorwulff committed:
Code stan changes due to missing @var

igorwulff committed:
Merge branch '2.4-develop' into patch-7

igorwulff commented:
@magento run all tests

igorwulff commented:
@magento run WebAPI Tests

igorwulff commented:
Hi @Den4ik can you clarify for me if the issues in the tests are related/caused by the changes in this PR?

igorwulff committed:
Add fix for xml loaded properties

Some object utilize xml properies to load more data and parameters into them, like the indexers through the constructor.
Checking if the empty item is just an instanceof \Magento\Framework\Model\AbstractModel might be enough to circumvent this potential problem.

igorwulff commented:
@magento run all tests
